pyexchange/kraken.py

```diff
@@ -0,0 +1,301 @@
+                orders.append(Order.from_our_order(order_id, order))
+
+        return orders
+
+    def place_order(self, pair: str, is_sell: bool, price: Wad, amount: Wad) -> str:
+        assert(isinstance(pair, str))
+        assert(isinstance(is_sell, bool))
+        assert(isinstance(price, Wad))
+        assert(isinstance(amount, Wad))
+
+        data = {
+            'pair': pair,
+            'type': 'sell' if is_sell else "buy",
+            'ordertype': 'limit',
+            "volume": str(amount),
+            "price": str(price)
+        }
+
+        self.logger.info(f"Placing order ({data['type']}, amount {data['volume']} of {pair},"
+                         f" price {data['price']})...")
+
+        result = self._http_authenticated("POST", "/0/private/AddOrder", data)
+
+        return result['txid'][0]
+
+    def cancel_order(self, order_id: str) -> bool:
+        assert(isinstance(order_id, str))
+
+        self.logger.info(f"Cancelling order #{order_id}...")
+        self._http_authenticated("POST", "/0/private/CancelOrder", {'txid': order_id})
+
+        return True
+
+    def get_trades(self, pair: str, page_number: int = 1) -> List[Trade]:
+        assert(isinstance(pair, str))
+        assert(isinstance(page_number, int))
+        assert(page_number == 1)
+
+        result = self._http_authenticated("POST", f"/0/private/TradesHistory", {})
+
+        trades = []
+
+        for trade_id, trade in result['trades'].items():
+            if trade['pair'] == pair and trade['ordertype'] == 'limit':
+                trades.append(Trade.from_our_trade(trade_id, trade))
+
+        return trades
+
+    def get_all_trades(self, pair: str, page_number: int = 1) -> List[Trade]:
+        assert(isinstance(pair, str))
+        assert(isinstance(page_number, int))
+
+        result = self._http_unauthenticated("POST", f"/0/public/Trades?pair={pair}", {})
+
+        return list(map(lambda item: Trade.from_all_response(pair, item), result[pair]))
+
+    def _http_authenticated(self, method: str, resource: str, body: dict):
+        assert(isinstance(method, str))
+        assert(isinstance(resource, str))
+        assert(isinstance(body, dict) or (body is None))
+
+        body['nonce'] = int(1000 * time.time())
+        postdata = urlencode(body)
+        encoded = (str(body['nonce']) + postdata).encode()
+        message = resource.encode() + hashlib.sha256(encoded).digest()
+        signature = hmac.new(base64.b64decode(self.secret_key), message, hashlib.sha512)
+        sigdigest = base64.b64encode(signature.digest())
+
+        headers = {
+            'API-Key': self.api_key,
+            'API-Sign': sigdigest.decode()
+        }
+
+        return self._result(requests.request(method=method,
+                                             url=f"{self.api_server}{resource}",
+                                             data=body,
+                                             headers=headers,
```

Comment:
move headers block into here just like in CB ->
```
  return self._result(requests.request(method=method,
                                             url=f"{self.api_server}{resource}",
                                             data=data,
                                             headers={
                                                 'Content-Type': 'Application/JSON',
                                                 'CB-ACCESS-SIGN': signature_b64,
                                                 'CB-ACCESS-TIMESTAMP': timestamp,
                                                 'CB-ACCESS-KEY': self.api_key,
                                                 'CB-ACCESS-PASSPHRASE': self.password
                                             },
                                             timeout=self.timeout))
```

Reply:
done